File: CodingSamples/DotNet/Advanced/Networking/TcpClientTest/DemoApp1/Models/ShopModel.cs
using System.Net.Sockets;

namespace DemoApp.Models;

public class ShopModel(string remote)
{
    public async ValueTask<ItemInfo> FetchItemInfoAsync(string name)
    {
        //Step 1
        var connection = new TcpClient(remote, 4000);
        //Step 2
        var channel = connection.GetStream();
        using var reader = new StreamReader(channel);
        using var writer = new StreamWriter(channel) { AutoFlush = true };
        await reader.ReadLineAsync(); //receive greeting message
        await writer.WriteLineAsync(name); //send item name
        string message = await reader.ReadLineAsync(); //receive item info
        //Step 3
        connection.Close();
        if(message != null)
            return ItemInfo.Parse(message);
        return default; //new ItemInfo()
    }
}

File: CodingSamples/Java/Advanced/Networking/TCPClientTest/demo1/src/main/java/app/models/ShopModel.java
package app.models;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

public class ShopModel {
    
    private String server;

    public ShopModel(String remote) {
        server = remote;
    }

    public ItemInfo fetchItemInfo(String name) throws IOException {
        //step 1
        var connection = new Socket(server, 4000);
        //step 2
        var receiver = connection.getInputStream();
        var reader = new BufferedReader(new InputStreamReader(receiver));
        var sender = connection.getOutputStream();
        var writer = new PrintWriter(sender, true);
        reader.readLine(); //read greeting message
        writer.println(name);
        String message = reader.readLine();
        writer.close();
        reader.close();
        //Step 3
        connection.close();
        if(message != null)
            return ItemInfo.parse(message);
        return new ItemInfo(0, 0);
    }
}
